subsys/net/lib/dns/dns_sd.c

```diff
@@ -0,0 +1,786 @@
+		/* RFC 6763 Section 4.1.1 */
+		if (instance[i] <= 0x1f ||
+		    instance[i] == 0x7f) {
+			LOG_DBG(
+				"instance '%s' contains illegal byte 0x%02x",
+				instance, instance[i]);
+			return false;
+		}
+	}
+
+	return instance_size;
+}
+
+static bool service_is_valid(const char *service)
+{
+	size_t service_size;
+
+	if (service == NULL) {
+		LOG_DBG("label is NULL");
+		return false;
+	}
+
+	service_size = strlen(service);
+	if (service_size < DNS_SD_SERVICE_MIN_SIZE) {
+		LOG_DBG("label '%s' is too small (%zu, min: %u)",
+			service, service_size, DNS_SD_SERVICE_MIN_SIZE);
+		return false;
+	}
+
+	if (service_size > DNS_SD_SERVICE_MAX_SIZE) {
+		LOG_DBG("label '%s' is too big (%zu, max: %u)",
+			service, service_size, DNS_SD_SERVICE_MAX_SIZE);
+		return false;
+	}
+
+	if (service[0] != DNS_SD_SERVICE_PREFIX) {
+		LOG_DBG("service '%s' invalid (no leading underscore)",
+			service);
+		return false;
+	}
+
+	if (!label_is_valid(&service[1], service_size - 1)) {
+		LOG_DBG("service '%s' contains invalid characters",
+			service);
+		return false;
+	}
+
+	return service_size;
+}
+
+static bool proto_is_valid(const char *proto)
+{
+	size_t proto_size;
+
+	if (proto == NULL) {
+		LOG_DBG("label is NULL");
+		return false;
+	}
+
+	proto_size = strlen(proto);
+	if (proto_size != DNS_SD_PROTO_SIZE) {
+		LOG_DBG("label '%s' wrong size (%zu, exp: %u)",
+			proto, proto_size, DNS_SD_PROTO_SIZE);
+		return false;
+	}
+
+	if (!(strncasecmp("_tcp", proto, DNS_SD_PROTO_SIZE) == 0 ||
+	      strncasecmp("_udp", proto, DNS_SD_PROTO_SIZE) == 0)) {
+		/* RFC 1034 Section 3.1 */
+		LOG_DBG("proto '%s' is invalid (not _tcp or _udp)",
+			proto);
+		return false;
+	}
+
+	return proto_size;
+}
+
+static bool domain_is_valid(const char *domain)
+{
+	size_t domain_size;
+
+	if (domain == NULL) {
+		LOG_DBG("label is NULL");
+		return false;
+	}
+
+	domain_size = strlen(domain);
+	if (domain_size < DNS_SD_DOMAIN_MIN_SIZE) {
+		LOG_DBG("label '%s' is too small (%zu, min: %u)",
+			domain, domain_size, DNS_SD_DOMAIN_MIN_SIZE);
+		return false;
+	}
+
+	if (domain_size > DNS_SD_DOMAIN_MAX_SIZE) {
+		LOG_DBG("label '%s' is too big (%zu, max: %u)",
+			domain, domain_size, DNS_SD_DOMAIN_MAX_SIZE);
+		return false;
+	}
+
+	if (!label_is_valid(domain, domain_size)) {
+		LOG_DBG("domain '%s' contains invalid characters",
+			domain);
+		return false;
+	}
+
+	return domain_size;
+}
+
+/**
+ * Check DNS SD Record for validity
+ *
+ * Our records are in the form <Instance>.<Service>.<Proto>.<Domain>
+ *
+ * Currently, <Subdomain>.<Domain> services are not supported.
+ */
+bool dns_sd_rec_is_valid(const struct dns_sd_rec *inst)
+{
+	return true
+	       && inst != NULL
+	       && instance_is_valid(inst->instance)
+	       && service_is_valid(inst->service)
+	       && proto_is_valid(inst->proto)
+	       && domain_is_valid(inst->domain)
+	       && inst->text != NULL
+	       && inst->port > 0
+	;
+}
+
+int add_a_record(const struct dns_sd_rec *inst, uint32_t ttl,
```

Comment:
Either not namespaced or should be static.

Reply:
They are static when not running tests.

Reply:
To be honest, I would have preferred to use the "unit" type in these tests, but due to some external conflict I could not. Had I been able to, I would have just included the .c file in the test.

Reply:
> Had I been able to, I would have just included the .c file in the test.

Including the .c file could be a nice solution here, what was the issue preventing that?

Reply:
This was the conflict

```
/usr/lib/ccache/cc -DZTEST_UNITTEST -I/home/chrisfriedt/workspace/zephyrproject/zephyr/subsys/testsuite/ztest/include -I/home/chrisfriedt/workspace/zephyrproject/zephyr/subsys/testsuite/include -I/home/chrisfriedt/workspace/zephyrproject/zephyr/include -I/home/chrisfriedt/workspace/zephyrproject/zephyr/. -Wall -I /home/chrisfriedt/workspace/zephyrproject/zephyr/build/include/generated -MD -MT CMakeFiles/testbinary.dir/src/main.c.o -MF CMakeFiles/testbinary.dir/src/main.c.o.d -o CMakeFiles/testbinary.dir/src/main.c.o   -c /home/chrisfriedt/workspace/zephyrproject/zephyr/tests/net/lib/dns_sd/src/main.c
In file included from /home/chrisfriedt/workspace/zephyrproject/zephyr/tests/net/lib/dns_sd/src/../../../../../subsys/net/lib/dns/dns_pack.h:10,
                 from /home/chrisfriedt/workspace/zephyrproject/zephyr/tests/net/lib/dns_sd/src/main.c:12:
/home/chrisfriedt/workspace/zephyrproject/zephyr/include/net/net_ip.h:1530:10: fatal error: syscalls/net_ip.h: No such file or directory
 #include <syscalls/net_ip.h>
          ^~~~~~~~~~~~~~~~~~~
compilation terminated.
```

Added this in `<net/net_ip.h>`
```diff
diff --git a/include/net/net_ip.h b/include/net/net_ip.h
index 22509a74a8..3299e27600 100644
--- a/include/net/net_ip.h
+++ b/include/net/net_ip.h
@@ -1527,7 +1527,9 @@ const char *net_family2str(sa_family_t family);
 }
 #endif
 
+#ifndef ZTEST_UNITTEST
 #include <syscalls/net_ip.h>
+#endif /* ZTEST_UNITTEST */
 
 /**
  * @}
```

And then I get this error:
```
FAILED: CMakeFiles/testbinary.dir/src/main.c.o 
/usr/lib/ccache/cc -DZTEST_UNITTEST -I/home/chrisfriedt/workspace/zephyrproject/zephyr/subsys/testsuite/ztest/include -I/home/chrisfriedt/workspace/zephyrproject/zephyr/subsys/testsuite/include -I/home/chrisfriedt/workspace/zephyrproject/zephyr/include -I/home/chrisfriedt/workspace/zephyrproject/zephyr/. -Wall -I /home/chrisfriedt/workspace/zephyrproject/zephyr/build/include/generated -MD -MT CMakeFiles/testbinary.dir/src/main.c.o -MF CMakeFiles/testbinary.dir/src/main.c.o.d -o CMakeFiles/testbinary.dir/src/main.c.o   -c /home/chrisfriedt/workspace/zephyrproject/zephyr/tests/net/lib/dns_sd/src/main.c
In file included from /home/chrisfriedt/workspace/zephyrproject/zephyr/include/net/net_ip.h:27,
                 from /home/chrisfriedt/workspace/zephyrproject/zephyr/tests/net/lib/dns_sd/src/../../../../../subsys/net/lib/dns/dns_pack.h:10,
                 from /home/chrisfriedt/workspace/zephyrproject/zephyr/tests/net/lib/dns_sd/src/main.c:12:
/home/chrisfriedt/workspace/zephyrproject/zephyr/include/sys/byteorder.h:20:36: error: expected declaration specifiers or ‘...’ before ‘(’ token
 #define __bswap_16(x) ((uint16_t) ((((x) >> 8) & 0xff) | (((x) & 0xff) << 8)))
                                    ^
/home/chrisfriedt/workspace/zephyrproject/zephyr/include/sys/byteorder.h:24:36: error: expected declaration specifiers or ‘...’ before ‘(’ token
 #define __bswap_32(x) ((uint32_t) ((((x) >> 24) & 0xff) | \
                                    ^
/home/chrisfriedt/workspace/zephyrproject/zephyr/include/sys/byteorder.h:34:36: error: expected declaration specifiers or ‘...’ before ‘(’ token
 #define __bswap_64(x) ((uint64_t) ((((x) >> 56) & 0xff) | \
                                    ^
```

Reply:
This is my diff to convert it to `type: unit` if you'd like to give it a shot.
I'm developing on an Aarch64 Chromebook atm, which is the reason for the change in `subsys/testsuite/unittest.cmake`.

The unit test framework also complained about sign for the argument passed to `strncpy()`, which is why that was changed.

I've not found a good way to add include paths with `type: unit`. Hence, all of the `../../../` in the test file. 

[convert-to-unit-test.patch.txt](https://github.com/zephyrproject-rtos/zephyr/files/5440410/convert-to-unit-test.patch.txt)

Reply:
Difficult to know what goes wrong in your out-of-tree testing system, I feel your pain but we should concentrate to get things working "cleanly" on in-tree code and with current ztest framework.

I have some doubts whether the changes in unittest.cmake would be accepted, but you could of course try to send a patch for that.

I do not get why we would need these kind of paths in test files, 
`+#include "../../../../../subsys/net/lib/dns/dns_pack.h"

We could keep things simple actually. You could for example define the internal functions as static if not unit testing them, and just make them non-static if `CONFIG_NET_TEST` is set which allows calling them from unit tests. I think this is the simplest way and we do not need to start to do some voodoo magic here. The ppp tests in `tests/net/ppp` are calling things similar way the functions in `drivers/net/ppp.c` driver.

Reply:
The reason all of the `../../../` are used (as I [wrote previously](https://github.com/zephyrproject-rtos/zephyr/pull/29179/files#r512148281)) is that I've not found the right way to include paths with `type: unit`. 

Originally, I had used this to add include directories, but that does not work with `type: unit` for some reason. 
```cmake
target_include_directories(app PRIVATE ${ZEPHYR_BASE}/subsys/net/lib/dns)
```

Other than that, it seems that my cross toolchain has some preprocessor conflict with Zephyr headers. I have not yet tried compiling the `type: unit` version with the Zephyr SDK. Am I missing something about specifying include directories for `type: unit` test builds?